### PR TITLE
dev-form Pull Request

### PR DESCRIPTION
**Pull request by FSolM**
Project's form implementation 📋

### CODE REVIEW

**Comment by FSolM** [comment]
I know that I have a `console.log` in the `books.js` reducer file, it's just for reference that the `switch` is working as intended and will delete it in the next Milestone

**Comment by FSolM** [comment]
The form "does something", hence the `console.log` statement, check the browser's log. Will address the initial statement part

**Comment by bolah2009** [comment]
> The form "does something", hence the `console.log` statement, check the browser's log. Will address the initial statement part

Please go back to the requirements (https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/44896828/) and make the book form functional. 

- We should be able to add books. Books added are meant to be rendered on the page. We have nothing to do with console.log (remove it).
- We should be able to remove books. Books rendered on the page should have a functional remove button, that deletes a book from the state and removes it from the page.

File: src/components/BooksForm.js [hunk 1]
```diff
@@ -1,35 +1,57 @@
...
+    setTitle(e.target.value);
+  };
+
+  const handleCategoryChange = (e) => {
+    setCategory(e.target.value);
+  };
+
+  const handleSubmit = () => {
+    if (title === '') {
+      alert("You shouldn't leave book names blank!");
```
Comment (bolah2009) [inline src/components/BooksForm.js]: [Optional]

I would recommend not using `alert` here. See [this SO discussion](https://stackoverflow.com/questions/20359807/why-shouldnt-i-use-alert-in-javascript)

File: src/components/BooksForm.js [hunk 2]
```diff
@@ -1,35 +1,57 @@
...
+      return;
+    }
+    const book = {
+      title: title,
+      category: category
+    };
+    dispatch(addBook(book));
+    document.getElementById('title').value = '';
+    document.getElementById('category').value = categories[0];
```
Comment (bolah2009) [inline src/components/BooksForm.js]: What are you doing here? :thinking: 

Even if you changed the inputs the current `state` is what will be computed. Why not reset the `state` instead?

image

See illustration from your app: The inputs are showing different values from what is been computed.

solerror


I would suggest the following:
Set the input values with the `state`, so when the `state` is reset to the initial value, the inputs are also reset.

File: src/components/BooksForm.js [hunk 3]
```diff
@@ -1,35 +1,57 @@
...
+  return (
+    <div className = 'booksform-container'>
+      <label htmlFor = 'title'>Title</label>
+      <input type = 'text' id = 'title' onChange = {(e) => { handleTitleChange(e) }} />
+      <label htmlFor = 'category'>Category</label>
+      <select id = 'category' onChange = {(e) => { handleCategoryChange(e) }}>
+        {renderOptions()}
+      </select>
+      <button onClick = {() => { handleSubmit() }}>Send!</button>
+    </div>
```
Comment (bolah2009) [inline src/components/BooksForm.js]: Why not wrap this in a `form` element and use the form `onSubmit` event listener?

File: src/components/BooksForm.js [hunk 4]
```diff
@@ -1,35 +1,57 @@
...
+
+  const renderOptions = () => {
     let options = [];
-    for (let option in this.categories) {
-      options.push(<option value = {option}>{option}</option>)
+    for (let i = 0; i < categories.length; i += 1) {
+      options.push(<option key = {i} value = {categories[i]}>{categories[i]}</option>);
     }
     return options;
-  }
```
Comment (bolah2009) [inline src/components/BooksForm.js]: [Optional]

We could use `map` here instead:

```jsx
  const renderOptions = categories.map(category => (
    <option key={category} value={category}>
      {category}
    </option>
  ));
```

File: src/components/BooksList.js [hunk 5]
```diff
@@ -1,35 +1,40 @@
...
-      // Calls Book component, renders values
-    return books;
+function BooksList(props) {
+  const renderBooks = () => {
+    let render = [];
+    props.books.bookList.map((book) => {
+      return render.push(<Book key = {book.id} id = {book.id} title = {book.title} category = {book.category} />);
+    })
+    return render;
   }
```
Comment (bolah2009) [inline src/components/BooksList.js]: See the previous comment on how to use `map` effectively,

File: src/reducers/books.js [hunk 6]
```diff
@@ -1,16 +1,12 @@
-const initialState = {
-  
-};
+const initialState = { bookList: [] };
```
Comment (bolah2009) [inline src/reducers/books.js]: Let's add example books to initial state object in the `src/index.js`

image

See the previous review comment on this,

File: src/components/BooksForm.js (hunk 1 above)
Comment (FSolM) [reply src/components/BooksForm.js]: it's just a temporary measure until I find something more appropriate

File: src/components/BooksForm.js (hunk 3 above)
Comment (FSolM) [reply src/components/BooksForm.js]: eh, I don't like working with `form` if I'm not implementing PHP scripts

File: src/reducers/books.js (hunk 6 above)
Comment (FSolM) [reply src/reducers/books.js]: seems redundant, but alright

File: src/components/BooksForm.js (hunk 4 above)
Comment (FSolM) [reply src/components/BooksForm.js]: I'm more of a traditionalist and prefer my `for` loops

File: src/components/BooksForm.js (hunk 2 above)
Comment (FSolM) [reply src/components/BooksForm.js]: hmm, thanks for spotting that, forgot about that `state`'s state

File: src/components/BooksList.js (hunk 5 above)
Comment (FSolM) [reply src/components/BooksList.js]: "using `map` effectively"? it just looks like a `map` function to me

**Comment by FSolM** [comment]
Found this [post](https://stackoverflow.com/questions/36619093/why-do-i-get-reducer-returned-undefined-during-initialization-despite-pr) explaining why supplying the `initialState` through the `createStore` declaration is not enough

File: src/components/BooksForm.js (hunk 3 above)
Comment (bolah2009) [reply src/components/BooksForm.js]: Hi, sorry should have tagged this with `optional` tag. But it makes the validation easier using HTML native form validation. Submit would fire until form inputs are valid.

File: src/components/BooksList.js (hunk 5 above)
Comment (bolah2009) [reply src/components/BooksList.js]: This is really not an effective way to use `map`. Do you know that `map` function returns an array of its own? Why would you create an empty array and push components into it using `map`? You could have done this with `forEach` or `for` loop.

The whole of these
```jsx
  const renderBooks = () => {
    let render = [];
    props.books.bookList.map((book) => {
      return render.push(<Book key = {book.id} id = {book.id} title = {book.title} category = {book.category} />);
    })
    return render;
  }
```
Could be written like this:
```jsx
  const renderBooks = props.books.bookList.map((book) => (
    <Book key = {book.id} id = {book.id} title = {book.title} category = {book.category} />
  );
```


Also, this is an optional requirement too. So you are free to ignore these suggestions.

**Comment by FSolM** [comment]
well, I have watched apps that behave similar, so I didn't pay much mind to that; but you're right, it's asthetically unpleasant, will check it out

**Pull request merged**